Fix totp registration workflow with broken authenticators

Firstyear opened the pull request:
Fixes #501 #506 - This changes the workflow in totp registration, so that we can detect broken authenticators. When we detect these, we request consent from the user to proceed in allowing this registration to proceed for the account. 

- [ x ] cargo fmt has been run
- [ ] cargo clippy has been run
- [ x ] cargo test has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

Firstyear commented:
cc @cuberoot74088

yaleman reviewed (approved):
This works for me in prod, thanks :)

Firstyear commented:
I'll let @cuberoot74088 have a look in the next ~24 hours, then I'll merge it. Thanks for always reviewing @yaleman

cuberoot74088 commented:
@Firstyear I will give it a try later today. Any thoughts on auto-magic downgrade vs. new user request with selected algo?

Firstyear commented:
> @Firstyear I will give it a try later today. Any thoughts on auto-magic downgrade vs. new user request with selected algo?

Many :) there is a reason I implemented this the way I did (request to downgrade after invalid authenticator).

The main reason is that we want the best defaults whenever possible. When someone writes documentation, or a blog, they'll say "kanidm account credentials register_totp". Now all it takes is one blogger, who hits this issue to say "need to add --algo=sha1" for this to work. People will copy paste it, and they'll have a different experience. People may question the need for the algo selection etc. They'll use sha1 without a consideration of the consequences (and why should they know those consequences? Not everyone is an authentication specialist, or knows the ins/outs of authenticator apps)

The intent here is to reduce options to guide people to a best default case, without the risk of weaking others. By having this as "auto-magic", we get:

* Strong defaults, for the majority of users, and less confusion because less options
* Clear, direct warnings for users who *are* affected by this issue, and that they have to consent to the downgrade
* An ability in the future for security policy to control if we answer this query allowing sha1 or not at all. 

So I've implemented the automagic method, where if you scan, get a sha1 authenticator, we warn you, and you have to consent to proceed. I don't want to be in a position where users have to "become experts" on a topic that isn't their area of expertise just to enable 2fa.

cuberoot74088 commented:
Hi @Firstyear,
thanks for taking the time and sharing your thoughts! :+1: 

I think your points are very valid, especially reducing the complexity (more options etc.) and demanding the expertise. I also see another benefit in the downgrade workflow, as the user does not need to scan another code.

I can also confirm, it is working:

> Issuer: demo_user2@example.com
> Algorithm: SHA256
> Period/Step: 30
> Secret: K6VLL6GXBJQBQ
>
> Enter a TOTP from your authenticator to complete registration:
> TOTP: 635291
> 
> ⚠️  WARNING - It appears your authenticator app may be broken ⚠️  
>  The TOTP authenticator you are using is forcing the use of SHA1
> 
>  -- If you accept this risk, and wish to proceed, type 'I am sure' 
>  -- Otherwise press ENTER to cancel this operation
> 
> Are you sure: I am sure
> TOTP registration success.


Another comment is the input loop. Maybe 3 or 10 times would be enough? Seems like I can try this forever

> Incorrect TOTP code - try again
> TOTP: 3
> 
> 4Incorrect TOTP code - try again
> TOTP: 

Again, huge thanks for the quick fix and the explanations :1st_place_medal:

cuberoot74088 commented on kanidm_tools/src/cli/account.rs:
```diff
@@ -189,37 +190,81 @@ impl AccountOpt {
+                                    break;
+                                };
+
+                                if confirm_input.trim() == "I am sure" {
```
Should that be case sensitive?

yaleman replied:
I'd argue yes - you really want them to type it properly, or copy-paste it like every one of these things I come across 😆

Firstyear replied:
I think even just at a point they type 'i am sure' that's still strong enough "intent" that insensitive here is probably fine. We just need it to be a deliberate and intended action, rather than an accidental accept.

cuberoot74088 commented on kanidm_tools/src/cli/account.rs:
```diff
@@ -189,37 +190,81 @@ impl AccountOpt {
+                            }
+                            Err(ClientError::TotpVerifyFailed(_, _)) => {
+                                eprintln!("Incorrect TOTP code - try again");
+                                continue;
```
As far as I have seen and tried, this will loop over and over again. Do we want this? Maybe limit the maximum tries?

Firstyear replied:
I think a limit here is a good idea. :)

Firstyear commented:
Applied @cuberoot74088's feedback :)
